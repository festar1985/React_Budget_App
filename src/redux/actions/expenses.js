import { push, onValue, remove, ref, update } from "firebase/database";
import { myDatabase, databaseReference } from "../../firebase/firebase";

//Actions

export const addExpense = (expenses) => ({
  type: "ADD_EXPENSE",
  expenses,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "Description not available",
      category = "Others",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, category, note, amount, createdAt };

    push(databaseReference, expense).then((reference) => {
      dispatch(
        addExpense({
          id: reference.key,
          ...expense,
        })
      );
    });
  };
};

export const removeExpense = (id = "") => ({ type: "REMOVE_EXPENSE", id });

export const startRemoveExpense = (id) => {
  return async (dispatch) => {
    await remove(ref(myDatabase, id));
    dispatch(removeExpense(id));
  };
};

export const modifyExpense = (id, updates) => ({
  type: "MODIFY_EXPENSE",
  id,
  updates,
});

export const startModifyExpense = (id, updates) => {
  return async (dispatch) => {
    await update(ref(myDatabase, id), updates);
    dispatch(modifyExpense(id, updates));
  };
};

export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    onValue(
      databaseReference,
      (snapshot) => {
        const expenses = [];
        Object.entries(snapshot.val()).forEach(([key, val]) => {
          const expense = {
            id: key,
            ...val,
          };
          expenses.push(expense);
        });

        dispatch(setExpenses([...expenses]));
      },
      { onlyOnce: true }
    );
  };
};
