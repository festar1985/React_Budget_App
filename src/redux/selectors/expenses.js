const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((element) => {
      const startDateMatch =
        typeof startDate != "number" || element.createdAt >= startDate;
      const endDateMatch =
        typeof startDate != "number" || element.createdAt <= endDate;
      const textMatch = element.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
