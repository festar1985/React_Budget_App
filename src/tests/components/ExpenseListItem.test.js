import React from "react";
import ExpensesListItem from "../../components/ExpenseListItem";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";
import { expenses } from "../fixture/expenses";

Enzyme.configure({ adapter: new Adapter() });

test("should render an expense list item when i give an example", () => {
  const wrapper = shallow(<ExpensesListItem {...expenses[0]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
