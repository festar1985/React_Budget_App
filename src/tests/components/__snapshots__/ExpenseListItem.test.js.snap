// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense list item when i give an example 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p />
  <p>
    30
  </p>
  <p>
    0
  </p>
</div>
`;
